### PR TITLE
Implement ETHStaticATokens and add a rewards controller framework

### PR DESCRIPTION
**Pull request by JorikSchellekens**
This pr needs to be rebased on the `token_bridge` branch.

Tests not implemented yet

### CODE REVIEW

File: contracts/src/starknet/rewaave/math/wad_ray_math.cairo [hunk 1]
```diff
@@ -0,0 +1,218 @@
+from starkware.cairo.common.uint256 import (
```
Comment (achab) [inline contracts/src/starknet/rewaave/math/wad_ray_math.cairo:1]: from where does this code come from? should we test it?

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+func mint{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        recipient : felt, amount : Uint256):
+    Ownable_only_owner()
+    claimable_before_token_transfer(0, recipient)
+    ERC20_mint(recipient, amount)
+    return ()
+end
+
+@external
+func claimRewards{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```
Comment (achab) [inline contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:134]: general question: shouldn't we use snake case for cairo?

File: contracts/src/starknet/rewaave/math/wad_ray_math.cairo (hunk 1 above)
Comment (JorikSchellekens) [reply contracts/src/starknet/rewaave/math/wad_ray_math.cairo:1]: This is a direct tranpsilation of the implementation in Solidity. There's already a ticket about testing it in the board (added it after the meeting). This is the original source: https://github.com/aave/protocol-v2/tree/1d80f9a860b5d389fbae9034338c01fac6f15aa2/contracts/protocol/libraries/math

File: contracts/src/starknet/rewaave/rewAAVEController.cairo [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+struct storage_slot:
+    member word_1 : felt
+    member word_2 : felt
+    member word_3 : felt
+    member word_4 : felt
+end
+
+@contract_interface
+namespace IFossil:
```
[the remark this answers is not in the document]
Comment (JorikSchellekens) [reply contracts/src/starknet/rewaave/rewAAVEController.cairo:16]: Yes, ticket on the board but I think I'll just remove this now

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo (hunk 2 above)
Comment (JorikSchellekens) [reply contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:134]: Yes, will fix this everywhere

File: contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        end
+    else:
+        return ()
+    end
+end
+
+@external
+func get_accRewardsPerToken{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        ) -> (accRewardsPerToken : Uint256):
+    return claimable_get_accRewardsPerToken()
```
[the remark this answers is not in the document]
Comment (JorikSchellekens) [reply contracts/src/starknet/rewaave/tokens/ETHstaticAToken.cairo:165]: I addressed this on swapnil's pr https://github.com/aave-starknet-project/aave-starknet-bridge/pull/6#discussion_r841510850.

File: contracts/src/starknet/rewaave/tokens/claimable.cairo [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+from openzeppelin.token.erc20.library import ERC20_totalSupply, ERC20_balanceOf, ERC20_mint
+from openzeppelin.access.ownable import Ownable_initializer, Ownable_only_owner, Ownable_get_owner
+
+@storage_var
+func acc_rewards_per_token() -> (res : Uint256):
+end
+
+# user => accRewardsPerToken at last interaction (in RAYs)
+@storage_var
+func user_snapshot_rewards_per_token(user : felt) -> (acc_rewards_per_token : Uint256):
```
Comment (achab) [inline contracts/src/starknet/rewaave/tokens/claimable.cairo:19]: Should we also update this mapping using L1 transfers of StaticAToken?

**Review by achab — APPROVED** [review]

**Pull request merged**